lib/attrsets.nix

```diff
@@ -355,6 +372,26 @@ rec {
   mapAttrsToList = f: attrs:
     map (name: f name attrs.${name}) (attrNames attrs);
 
+  /* Map each attribute in the given set into multiple `nameValurPair`s
+     and return the result in a new attribute set.
+
+     Type:
+       mapAttrsMulti ::
+         (String -> a -> [{ name :: String, value :: b }])
+         -> AttrSet
+         -> AttrSet
+
+     Example:
+       mapAttrsMulti
+         (name: value: [
+           (nameValuePair name value)
+           (nameValuePair (name + value) value)
+         ])
+         { x = "a"; y = "b"; }
+       => { x = "a"; xa = "a"; y = "b"; yb = "b"; }
+  */
+  mapAttrsMulti = f: flip pipe [ (mapAttrsToList f) concatLists listToAttrs ];
+
```

Comment:
This function is related to `concatMap`.

Let's first generalize `concatMap`. I'll use `List a` instead of `[a]`.

```haskell
concatMap :: (a -> List b) -> List a -> List b

-- generalize List -->

bind :: (a -> f b) -> f a -> f b

-- specialize f to Attrset -->

concatMapAttrs :: (a -> Attrset b) -> Attrset a -> Attrset b
```

This is a slightly different, but perhaps nicer interface.

```nix
# Example:
       concatMapAttrs
         (name: value: {
           ${name} = value; 
           "${name + value}" = value;
         })
         { x = "a"; y = "b"; }
       => { x = "a"; xa = "a"; y = "b"; yb = "b"; }
```

It'd be something like this, I think.

```
flip pipe [ (mapAttrs f) attrValues (foldl' merge {}) ]
```

I would think that the last function was already available somewhere.


I also think `joinAttrs :: (String -> String -> String) -> Attrset (Attrset a) -> Attrset a` would be a good function to have.
It could be generalized to `joinAttrsRecursiveUntil` as well, with `(List String -> String)`.

Well, the `join` ones are just a thought, but `concatMapAttrs` will make this PR a bit cleaner.

Reply:
I'm not sure if `bind` is possible with nix's type system, maybe we just need something like `foldMap`

Reply:
`bind` was just to illustrate the transformation. Might have been a bit too elaborate.